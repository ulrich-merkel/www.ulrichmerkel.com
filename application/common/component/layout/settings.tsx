/**
 * Es6 module for handling theming as component.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, Component, Fragment, ReactNode } from 'react';
import { connect } from 'react-redux';
import PubSub from 'pubsub-js';

import { PUBSUB_COLOR_SCHEME_CHANGE_MESSAGE } from '../../state/color-scheme/constants';
import { selectStateColorSchemeSelected } from '../../state/color-scheme/selector';
import { selectStateReducedMotionSelected } from '../../state/reduced-motion/selector';
import { PUBSUB_REDUCED_MOTION_CHANGE_MESSAGE } from '../../state/reduced-motion/constants';
import { AvailableColorSchemesType } from '../../state/color-scheme/types';
import { AvailableReducedMotionsType } from '../../state/reduced-motion/types';

type Props = {
    children?: ReactNode;
    colorSchemeSelected: AvailableColorSchemesType;
    reducedMotionSelected: AvailableReducedMotionsType;
};

/**
 * Apply theming and settings.
 *
 * @augments React.Component
 * @property {ReactElement} props.children - Component childs to be rendered
 * @property {string} props.colorSchemeSelected - Current selected color scheme
 * @property {string} props.reducedMotionSelected - Current selected reduced motion
 */
export class LayoutSettings extends Component<Props> {
    /**
     * Invoked once, only on the client (not on the server),
     * immediately after the initial rendering occurs.
     *
     * @returns {void}
     */
    componentDidMount(): void {
        const { colorSchemeSelected, reducedMotionSelected } = this.props;

        PubSub.publish(PUBSUB_COLOR_SCHEME_CHANGE_MESSAGE, colorSchemeSelected);
        PubSub.publish(
            PUBSUB_REDUCED_MOTION_CHANGE_MESSAGE,
            reducedMotionSelected
        );
    }

    /**
     * Invoked immediately after the component's updates are flushed to
     * the DOM. This method is not called for the initial render.
     *
     * @param {object} prevProps - The old component properties
     * @returns {void}
     */
    componentDidUpdate(prevProps: Props): void {
        const { colorSchemeSelected, reducedMotionSelected } = this.props;

        if (colorSchemeSelected !== prevProps.colorSchemeSelected) {
            PubSub.publish(
                PUBSUB_COLOR_SCHEME_CHANGE_MESSAGE,
                colorSchemeSelected
            );
        }
        if (reducedMotionSelected !== prevProps.reducedMotionSelected) {
            PubSub.publish(
                PUBSUB_REDUCED_MOTION_CHANGE_MESSAGE,
                reducedMotionSelected
            );
        }
    }

    /**
     * The required render function to return a single react child element.
     *
     * @returns {ReactElement} React component markup
     */
    render(): ReactNode {
        const { children } = this.props;

        return <Fragment>{children}</Fragment>;
    }
}

/**
 * The component will subscribe to Redux store updates. Any time it updates,
 * mapStateToProps will be called, Its result must be a plain object,
 * and it will be merged into the componentâ€™s props.
 *
 * @private
 * @param {object<string, *>} state - The current redux store state
 * @returns {object<string, *>} The mapped state properties
 */
function mapStateToProps(state) {
    return {
        colorSchemeSelected: selectStateColorSchemeSelected(state),
        reducedMotionSelected: selectStateReducedMotionSelected(state)
    };
}

/**
 * Connects a React component to a Redux store. It does not modify the
 * component class passed to it. Instead, it returns a new, connected component class.
 *
 * @type {ReactElement}
 */
export const LayoutSettingsConnected = connect(mapStateToProps)(LayoutSettings);
