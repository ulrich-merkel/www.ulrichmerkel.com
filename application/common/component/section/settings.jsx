/* eslint-disable immutable/no-mutation */
/**
 * Es6 module for React Component.
 * Section components combine modules and elements
 * to create a page section which could be easily
 * used in page components.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import * as React from 'react';
import PropTypes from 'prop-types';

import { SectionCommonGridSpaced } from './common/grid-spaced';
import { ModuleArticle } from '../module/article';

/**
 * Function representing a component to return a single react child element.
 *
 * @param {object} [props] - The current component props
 * @param {Array|string} [props.children] - The component dom node childs - usally an array of components, if there is only a single child it's a string
 * @param {object} [props.content={}] - The content config input
 * @param {boolean} [props.isDialog=false] - Whether the component text should be displayed in a dialog or not
 * @param {boolean} [props.isMain=false] - Whether the component text should be displayed as main article or not
 * @returns {ReactElement} React component markup
 */
function SectionSettings(props) {
    const { children, content, isDialog, isMain } = props;

    return (
        <SectionCommonGridSpaced>
            <ModuleArticle {...{ content, isMain, isDialog }}>
                {children}
            </ModuleArticle>
        </SectionCommonGridSpaced>
    );
}

/**
 * Validate props via React.PropTypes helpers.
 *
 * @static
 * @type {object}
 */
SectionSettings.propTypes = {
    children: PropTypes.node, // eslint-disable-line react/require-default-props
    content: PropTypes.objectOf(
        PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number,
            PropTypes.array,
            PropTypes.object
        ])
    ),
    isDialog: PropTypes.bool,
    isMain: PropTypes.bool
};

/**
 * Set defaults if props aren't available.
 *
 * @static
 * @type {object}
 */
SectionSettings.defaultProps = {
    content: {},
    isDialog: false,
    isMain: false
};

export { SectionSettings };
