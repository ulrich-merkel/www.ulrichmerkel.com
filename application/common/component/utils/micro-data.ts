/**
 * Es6 module for creating mirco data attributes.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isValidString } from '../../utils/string';

type ItemTypeAttributes = {
    itemScope: boolean;
    itemType: string;
};

/**
 * Get component properties for itemProp and itemType.
 *
 * @param {boolean} itemType - The itemType property if available
 * @returns {object|null} The component props for spreading
 */
export function getItemTypeAttributes(
    itemType?: string
): ItemTypeAttributes | null {
    return isValidString(itemType)
        ? {
              itemScope: true,
              itemType
          }
        : null;
}
