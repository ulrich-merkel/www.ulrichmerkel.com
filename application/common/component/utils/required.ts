/**
 * Es6 module for creating form input required attributes.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isBoolean } from 'lodash';

type RequiredAttributes = {
    'aria-required': boolean;
    required: boolean;
};

/**
 * Get component properties for required form input elements.
 *
 * @param {boolean} required - Whether the input element is required or not
 * @returns {object|null} The component props for spreading
 */
export function getRequiredAttributes(
    required?: boolean
): RequiredAttributes | null {
    return isBoolean(required) && required
        ? {
              'aria-required': true,
              required: true
          }
        : null;
}
