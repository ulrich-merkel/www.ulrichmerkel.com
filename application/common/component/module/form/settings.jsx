/* eslint-disable immutable/no-mutation */
/**
 * Es6 module for React Component.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import * as React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { toggleThemeSelected } from '../../../state/color-scheme/duck';
import { selectStateCsrfToken } from '../../../state/csrf/selector';
import { selectStateColorSchemeSelectedDarkMode } from '../../../state/color-scheme/selector';
import { toggleReducedMotionSelected } from '../../../state/reduced-motion/duck';
import { selectStateReducedMotionSelectedReduce } from '../../../state/reduced-motion/selector';
import { eventPreventDefault } from '../../../utils/event';
import { GridRow } from '../../grid/row';
import { GridCol } from '../../grid/col';
import { Toggle } from '../../element/toggle';
import { Fieldset } from '../../element/fieldset';
import { Form } from '../../element/form';
import { Legend } from '../../element/legend';

const noop = Function.prototype;

/**
 * Function representing a component to return a single react child element.
 *
 * @param {object} [props] - The current component props
 * @returns {ReactElement} React component markup
 */
export function ModuleFormSettings(props) {
    const {
        content,
        csrfToken,
        colorSchemeSelectedDarkMode,
        onToggleReducedMotionSelected,
        reducedMotionSelectedReduce,
        onToggleThemeSelected
    } = props;

    return (
        <Form
            action="/theme/"
            className="m-form--theme"
            id="m-form--theme"
            itemProp="potentialAction"
            itemScope
            itemType="http://schema.org/Action"
            onSubmit={eventPreventDefault}
        >
            <Fieldset>
                <Legend isVisuallyHidden>{content.legend}</Legend>

                <GridRow>
                    <GridCol cols={'6'}>
                        <Toggle
                            id="dark-mode"
                            label="Dark mode"
                            onChange={onToggleThemeSelected}
                            checked={colorSchemeSelectedDarkMode}
                        />
                    </GridCol>
                </GridRow>
                <GridRow>
                    <GridCol cols={'6'}>
                        <Toggle
                            id="reduced-motion-toggle"
                            label="Reduced motion"
                            checked={reducedMotionSelectedReduce}
                            onChange={onToggleReducedMotionSelected}
                        />
                    </GridCol>
                </GridRow>
                <input type="hidden" name="_csrf" value={csrfToken} />
            </Fieldset>
        </Form>
    );
}

/**
 * Validate props via React.PropTypes helpers.
 *
 * @static
 * @type {object}
 */
ModuleFormSettings.propTypes = {
    content: PropTypes.shape({
        legend: PropTypes.string
    }),
    colorSchemeSelectedDarkMode: PropTypes.bool,
    csrfToken: PropTypes.string,
    onToggleReducedMotionSelected: PropTypes.func,
    onToggleThemeSelected: PropTypes.func,
    reducedMotionSelectedReduce: PropTypes.bool
};

/**
 * Set defaults if props aren't available.
 *
 * @static
 * @type {object}
 * @see ModuleFormContact.propTypes
 */
ModuleFormSettings.defaultProps = {
    colorSchemeSelectedDarkMode: false,
    content: {},
    csrfToken: '',
    onToggleReducedMotionSelected: noop,
    onToggleThemeSelected: noop,
    reducedMotionSelectedReduce: false
};

/**
 * The component will subscribe to Redux store updates. Any time it updates,
 * mapStateToProps will be called, Its result must be a plain object,
 * and it will be merged into the component’s props.
 *
 * @private
 * @param {object.<*>} state - The redux store state
 * @returns {object}
 */
function mapStateToProps(state) {
    return {
        csrfToken: selectStateCsrfToken(state),
        colorSchemeSelectedDarkMode: selectStateColorSchemeSelectedDarkMode(
            state
        ),
        reducedMotionSelectedReduce: selectStateReducedMotionSelectedReduce(
            state
        )
    };
}

/**
 * If an object is passed, each function inside it will be assumed to
 * be a Redux action creator. An object with the same function names,
 * but with every action creator wrapped into a dispatch call so they
 * may be invoked directly, will be merged into the component’s props.
 * If a function is passed, it will be given dispatch.
 *
 * @private
 * @type {object<string, Function>}
 */
const mapDispatchToProps = {
    onToggleThemeSelected: toggleThemeSelected,
    onToggleReducedMotionSelected: toggleReducedMotionSelected
};

/**
 * Connects a React component to a Redux store. It does not modify the
 * component class passed to it. Instead, it returns a new, connected component class.
 */
export const ModuleFormSettingsConnected = connect(
    mapStateToProps,
    mapDispatchToProps
)(ModuleFormSettings);
