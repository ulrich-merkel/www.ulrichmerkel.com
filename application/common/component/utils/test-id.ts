/**
 * Es6 module for creating test id related attributes
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isValidString } from '../../utils/string';

type TestIdAttributes = {
    'data-testid': string;
};

/**
 * Get component properties for testing.
 *
 * @param {string} testId - The test id to be added
 * @returns {object|null} The component props for spreading
 */
export function getTestIdAttributes(testId?: string): TestIdAttributes | null {
    return isValidString(testId)
        ? {
              'data-testid': testId
          }
        : null;
}
