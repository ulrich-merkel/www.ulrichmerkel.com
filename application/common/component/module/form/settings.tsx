/* eslint-disable immutable/no-mutation */
/**
 * Es6 module for React Component.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, ChangeEvent, FunctionComponent } from 'react';
import { connect } from 'react-redux';
import { noop } from 'lodash';

import { toggleColorSchemeSelected } from '../../../state/color-scheme/duck';
import { selectStateCsrfToken } from '../../../state/csrf/selector';
import { selectStateColorSchemeSelectedDarkMode } from '../../../state/color-scheme/selector';
import { toggleReducedMotionSelected } from '../../../state/reduced-motion/duck';
import { selectStateReducedMotionSelectedReduce } from '../../../state/reduced-motion/selector';
import { eventPreventDefault } from '../../../utils/event';
import { GridRow } from '../../grid/row';
import { GridCol } from '../../grid/col';
import { Toggle } from '../../element/toggle';
import { Fieldset } from '../../element/fieldset';
import { Form } from '../../element/form';
import { Legend } from '../../element/legend';
import { getItemTypeAttributes } from '../../utils/micro-data';

type Props = {
    content: {
        legend: string;
    };
    csrfToken: '';
    colorSchemeSelectedDarkMode: boolean;
    onToggleReducedMotionSelected: (event: ChangeEvent) => void;
    reducedMotionSelectedReduce: boolean;
    onToggleThemeSelected: (event: ChangeEvent) => void;
};

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {object} [props] - The current component props
 * @returns {ReactElement} React component markup
 */
export const ModuleFormSettings: FunctionComponent<Props> = (props) => {
    const {
        content,
        csrfToken = '',
        colorSchemeSelectedDarkMode = false,
        onToggleReducedMotionSelected = noop,
        reducedMotionSelectedReduce = false,
        onToggleThemeSelected = noop
    } = props;

    return (
        <Form
            action="/settings/"
            className="m-form--settings"
            id="m-form--settings"
            onSubmit={eventPreventDefault}
            {...getItemTypeAttributes('http://schema.org/Action')}
        >
            <Fieldset>
                <Legend isVisuallyHidden>{content?.legend}</Legend>

                <GridRow>
                    <GridCol cols={12}>
                        <Toggle
                            id="dark-mode"
                            label="Dark mode"
                            onChange={onToggleThemeSelected}
                            checked={colorSchemeSelectedDarkMode}
                        />
                    </GridCol>
                </GridRow>
                <GridRow>
                    <GridCol cols={12}>
                        <Toggle
                            id="reduced-motion-toggle"
                            label="Reduced motion"
                            checked={reducedMotionSelectedReduce}
                            onChange={onToggleReducedMotionSelected}
                        />
                    </GridCol>
                </GridRow>
                <input type="hidden" name="_csrf" value={csrfToken} />
            </Fieldset>
        </Form>
    );
};

/**
 * The component will subscribe to Redux store updates. Any time it updates,
 * mapStateToProps will be called, Its result must be a plain object,
 * and it will be merged into the component’s props.
 *
 * @private
 * @param {object.<*>} state - The redux store state
 * @returns {object}
 */
function mapStateToProps(state) {
    return {
        csrfToken: selectStateCsrfToken(state),
        colorSchemeSelectedDarkMode: selectStateColorSchemeSelectedDarkMode(
            state
        ),
        reducedMotionSelectedReduce: selectStateReducedMotionSelectedReduce(
            state
        )
    };
}

/**
 * If an object is passed, each function inside it will be assumed to
 * be a Redux action creator. An object with the same function names,
 * but with every action creator wrapped into a dispatch call so they
 * may be invoked directly, will be merged into the component’s props.
 * If a function is passed, it will be given dispatch.
 *
 * @private
 * @type {object<string, Function>}
 */
const mapDispatchToProps = {
    onToggleThemeSelected: toggleColorSchemeSelected,
    onToggleReducedMotionSelected: toggleReducedMotionSelected
};

/**
 * Connects a React component to a Redux store. It does not modify the
 * component class passed to it. Instead, it returns a new, connected component class.
 */
export const ModuleFormSettingsConnected = connect(
    mapStateToProps,
    mapDispatchToProps
)(ModuleFormSettings);
