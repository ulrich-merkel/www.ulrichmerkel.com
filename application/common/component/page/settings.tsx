/**
 * Es6 module for React Component.
 * Page components combine section components to the
 * overall page layout.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, FunctionComponent } from 'react';
import Helmet from 'react-helmet';

import { addPageTracking } from '../decorator/add-page-tracking';
import { addContent } from '../decorator/add-content';
import { getContentSection } from '../../utils/content';
import { LayoutMain } from '../layout/main';
import { SectionSettings } from '../section/settings';
import { ModuleFormSettingsConnected } from '../module/form/settings';

type Props = {
    content: Record<string, unknown>;
    isDialog?: boolean;
};

/**
 * Function representing a component to return a single react child element.
 *
 * @function
 * @param {object} [props] - The current component props
 * @returns {ReactElement} React component markup
 */
const Page: FunctionComponent<Props> = (props) => {
    const { content, isDialog, children } = props;
    const contentSection = getContentSection(content);

    const pageThemeContent = (
        <SectionSettings content={contentSection('section1')} isMain isDialog>
            <ModuleFormSettingsConnected
                content={contentSection('formSettings')}
            />
            {children}
        </SectionSettings>
    );

    if (isDialog) {
        return pageThemeContent;
    }

    return (
        <LayoutMain>
            <Helmet {...contentSection('head')} />
            {pageThemeContent}
        </LayoutMain>
    );
};

export const PageSettings = addPageTracking(addContent('PageSettings')(Page));
