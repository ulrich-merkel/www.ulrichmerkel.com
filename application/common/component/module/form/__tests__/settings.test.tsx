import * as React from 'react';
import { render } from '../../../../../__tests__/utils/test-utils';
import { ModuleFormSettingsConnected } from '../settings';

describe('ModuleFormSettings', function fnDescribe() {
    const props = {
        content: {
            legend: 'Hello settings'
        },
        csrfToken: '',
        themeSelected: 'default',
        onchangeThemeSelectedDefault: jest.fn(),
        onchangeThemeSelectedGrey: jest.fn()
    };

    it('should render correctly', function fnIt() {
        const { asFragment } = render(
            <ModuleFormSettingsConnected {...props}>
                Module form children
            </ModuleFormSettingsConnected>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
