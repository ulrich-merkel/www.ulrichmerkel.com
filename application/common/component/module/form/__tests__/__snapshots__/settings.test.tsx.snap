// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleFormSettings should render correctly 1`] = `
<DocumentFragment>
  <form
    accept-charset="utf-8"
    action="/settings/"
    class="m-form m-form--settings"
    id="m-form--settings"
    itemprop="potentialAction"
    itemscope=""
    itemtype="http://schema.org/Action"
    method="post"
    novalidate=""
    role="form"
  >
    <fieldset
      class="c-fieldset m-form__fieldset"
    >
      <legend
        class="m-form__legend is-visually-hidden"
      >
        Hello settings
      </legend>
      <div
        class="l-grid__row"
      >
        <div
          class="l-grid__col--12"
        >
          <div
            class="c-toggle"
          >
            <input
              class="m-form__control--text c-toggle__input"
              id="dark-mode"
              name="dark-mode"
              placeholder=""
              type="checkbox"
              value=""
            />
            <label
              class="m-form__label c-toggle__label"
              for="dark-mode"
            >
              <span
                class="c-toggle__text"
              >
                Dark mode
              </span>
            </label>
          </div>
        </div>
      </div>
      <div
        class="l-grid__row"
      >
        <div
          class="l-grid__col--12"
        >
          <div
            class="c-toggle"
          >
            <input
              class="m-form__control--text c-toggle__input"
              id="reduced-motion-toggle"
              name="reduced-motion-toggle"
              placeholder=""
              type="checkbox"
              value=""
            />
            <label
              class="m-form__label c-toggle__label"
              for="reduced-motion-toggle"
            >
              <span
                class="c-toggle__text"
              >
                Reduced motion
              </span>
            </label>
          </div>
        </div>
      </div>
      <input
        name="_csrf"
        type="hidden"
        value="da39a3ee5e6b4b0d3255bfef95601890afd80709"
      />
    </fieldset>
  </form>
</DocumentFragment>
`;
